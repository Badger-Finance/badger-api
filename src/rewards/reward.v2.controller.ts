import { Network } from '@badger-dao/sdk';
import { Controller, Get, Inject, PathParams, QueryParams, UseCache } from '@tsed/common';
import { ContentType, Deprecated, Description, Hidden, Returns, Summary } from '@tsed/schema';

import { Chain } from '../chains/config/chain.config';

import { RewardsService } from './rewards.service';

import { AirdropMerkleClaim } from './interfaces/merkle-distributor.interface';
import { RewardMerkleClaimModel } from './interfaces/reward-merkle-claim-model.interface';
import { EmissionSchedule, RewardSchedulesByVaults } from './interfaces/reward-schedules-vault.interface';
import { RewardSchedulesByVaultModel } from './interfaces/reward-schedules-vault-model.interface';
import { RewardSchedulesByVaultsModel } from './interfaces/reward-schedules-vaults-model.interface';

@Deprecated()
@Controller('/reward')
export class RewardV2Controller {
  @Inject()
  rewardsService!: RewardsService;

  @Hidden()
  @Get('/bouncer/:address')
  @ContentType('json')
  async getBouncerProof(
    @PathParams('address') address: string,
    @QueryParams('chain') chain?: Network,
  ): Promise<AirdropMerkleClaim> {
    return this.rewardsService.getBouncerProof(Chain.getChain(chain), address);
  }

  @Get('/tree/:address')
  @ContentType('json')
  @Summary("Get an account's reward proof")
  @Description('Return user badger tree reward proof')
  @Returns(200, RewardMerkleClaimModel)
  @Returns(404).Description('User has no rewards proof available')
  async getBadgerTreeReward(
    @PathParams('address') address: string,
    @QueryParams('chain') chain?: Network,
  ): Promise<RewardMerkleClaimModel> {
    return this.rewardsService.getUserRewards(Chain.getChain(chain), address);
  }

  @UseCache()
  @Get('/schedules')
  @ContentType('json')
  @Summary('Get all token rewards emmited for all vaults on network')
  @Description('Return emission schedule list for all vaults')
  @Returns(200, RewardSchedulesByVaultsModel)
  async getRewardSchedulesVaultsList(
    @QueryParams('chain') chain?: Network,
    @QueryParams('active') active?: boolean,
  ): Promise<RewardSchedulesByVaults> {
    return this.rewardsService.rewardSchedulesVaultsList(Chain.getChain(chain), Boolean(active));
  }

  @UseCache()
  @Get('/schedules/:address')
  @ContentType('json')
  @Summary('Get all token rewards emmited for vault on network')
  @Description('Return emission schedule list for specified vault')
  @Returns(200, Array).Of(RewardSchedulesByVaultModel)
  @Returns(404).Description('Unknown vault')
  async getRewardListSchedulesForVault(
    @PathParams('address') address: string,
    @QueryParams('chain') chain?: Network,
    @QueryParams('active') active?: boolean,
  ): Promise<EmissionSchedule[]> {
    return this.rewardsService.rewardSchedulesByVault(Chain.getChain(chain), address, Boolean(active));
  }
}
