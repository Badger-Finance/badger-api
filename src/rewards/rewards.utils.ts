import { Network, Protocol } from '@badger-dao/sdk';
import { BigNumber } from 'ethers';

import { getBoostFile, getCachedAccount } from '../accounts/accounts.utils';
import { CachedValueSource } from '../aws/models/apy-snapshots.model';
import { VaultDefinitionModel } from '../aws/models/vault-definition.model';
import { getObject } from '../aws/s3.utils';
import { Chain } from '../chains/config/chain.config';
import { ONE_YEAR_SECONDS, REWARD_DATA } from '../config/constants';
import { TOKENS } from '../config/tokens.config';
import { getPrice } from '../prices/prices.utils';
import { createValueSource, ValueSource } from '../protocols/interfaces/value-source.interface';
import { BalancerStrategy } from '../protocols/strategies/balancer.strategy';
import { ConvexStrategy } from '../protocols/strategies/convex.strategy';
import { OxDaoStrategy } from '../protocols/strategies/oxdao.strategy';
import { QuickswapStrategy } from '../protocols/strategies/quickswap.strategy';
import { SushiswapStrategy } from '../protocols/strategies/sushiswap.strategy';
import { SwaprStrategy } from '../protocols/strategies/swapr.strategy';
import { UniswapStrategy } from '../protocols/strategies/uniswap.strategy';
import { getFullToken, tokenEmission } from '../tokens/tokens.utils';
import { getCachedVault, getVaultPerformance } from '../vaults/vaults.utils';
import { SourceType } from './enums/source-type.enum';
import { RewardMerkleDistribution } from './interfaces/merkle-distributor.interface';

export const DIGG_SHARE_PER_FRAGMENT = '222256308823765331027878635805365830922307440079959220679625904457';

export async function getTreeDistribution(chain: Chain): Promise<RewardMerkleDistribution | null> {
  try {
    const fileName = `badger-tree-${parseInt(chain.chainId, 16)}.json`;
    const rewardFile = await getObject(REWARD_DATA, fileName);
    return JSON.parse(rewardFile.toString('utf-8'));
  } catch (err) {
    console.error({ message: `Missing expected badger tree file for ${chain.network}`, err });
    return null;
  }
}

export async function getClaimableRewards(
  chain: Chain,
  chainUsers: string[],
  distribution: RewardMerkleDistribution,
  blockNumber: number,
): Promise<[string, [string[], BigNumber[]]][]> {
  const { rewards } = await chain.getSdk();
  const { badgerTree } = rewards;
  const requests = chainUsers.map(async (user): Promise<[string, [string[], BigNumber[]]]> => {
    const proof = distribution.claims[user];
    if (!proof) {
      return [user, [[], []]];
    }
    let attempt = 0;
    while (attempt < 3) {
      try {
        const result = await badgerTree.getClaimableFor(user, proof.tokens, proof.cumulativeAmounts, {
          blockTag: blockNumber,
        });
        return [user, result];
      } catch (err) {
        for (let i = 0; i < proof.tokens.length; i++) {
          const token = proof.tokens[i];
          const amount = await badgerTree.claimed(user, token);
          if (BigNumber.from(proof.cumulativeAmounts[i]).lt(amount)) proof.cumulativeAmounts[i] = amount.toString();
        }
        attempt++;
        // report a recurring issue for claimable
        if (attempt === 3) {
          console.error(err);
        }
      }
    }
    return [user, [[], []]];
  });
  return Promise.all(requests);
}

export async function getRewardEmission(chain: Chain, vault: VaultDefinitionModel): Promise<CachedValueSource[]> {
  const boostFile = await getBoostFile(chain);
  const sdk = await chain.getSdk();

  if (!sdk.rewards.hasRewardsLogger() || vault.depositToken === TOKENS.DIGG || !boostFile) {
    return [];
  }
  const { address } = vault;
  const cachedVault = await getCachedVault(chain, vault);

  if (address === TOKENS.BVECVX) {
    delete boostFile.multiplierData[address];
  }
  const boostRange = boostFile.multiplierData[address] ?? { min: 1, max: 1 };
  const activeSchedules = await sdk.rewards.loadActiveSchedules(address);

  // Badger controlled addresses are blacklisted from receiving rewards. We only dogfood on ETH
  let ignoredTVL = 0;
  if (chain.network === Network.Ethereum) {
    const blacklistedAccounts = await Promise.all([
      getCachedAccount(chain, '0xB65cef03b9B89f99517643226d76e286ee999e77'), // dev multisig
      getCachedAccount(chain, '0x86cbD0ce0c087b482782c181dA8d191De18C8275'), // tech ops multisig
      getCachedAccount(chain, '0x042B32Ac6b453485e357938bdC38e0340d4b9276'), // treasury ops multisig
      getCachedAccount(chain, '0xD0A7A8B98957b9CD3cFB9c0425AbE44551158e9e'), // treasury vault
      getCachedAccount(chain, '0xA9ed98B5Fb8428d68664f3C5027c62A10d45826b'), // treasury bveCVX voting multisig
    ]);
    ignoredTVL = blacklistedAccounts
      .map((a) => a.data[vault.address])
      .map((s) => (s ? s.value : 0))
      .reduce((total, value) => total + value, 0);
  }

  /**
   * Calculate rewards emission percentages:
   *   - P: Price of Token
   *   - A: Amount of Token Emitted
   *   - D: Duration of Token Emission in Seconds
   *   - S: Number of Seconds Per Year
   *   - L: Amount of Value Locked in Sett
   *
   * Reward APR is determined by the extrapolated one year value emitted
   * from the unlock schedues as a percentage of the current locked value.
   *
   * APR = (P * A * (S / D)) / L * 100
   *
   * This vaule is calculated for all tokens emitted for any given sett
   * and persisted against that sett. There is a 20 minute grace period for
   * emission that have since ended, and only the latest active emission
   * will be used for yield calcuation.
   */

  const emissionSources = [];
  for (const schedule of activeSchedules) {
    const tokenPrice = await getPrice(schedule.token);
    const token = await getFullToken(chain, schedule.token);

    const durationScalar = ONE_YEAR_SECONDS / (schedule.end - schedule.start);
    const yearlyEmission = tokenPrice.price * schedule.amount * durationScalar;
    const apr = (yearlyEmission / (cachedVault.value - ignoredTVL)) * 100;
    let proRataAPR = apr;
    if (cachedVault.boost.enabled && token.address === chain.getBadgerTokenAddress()) {
      const boostedAPR = (cachedVault.boost.weight / 10_000) * proRataAPR;
      proRataAPR = proRataAPR - boostedAPR;
      const boostedSource = createValueSource(`Boosted ${token.name}`, boostedAPR, boostRange);
      emissionSources.push(valueSourceToCachedValueSource(boostedSource, vault, tokenEmission(token, true)));
    }
    const proRataSource = createValueSource(`${token.name}`, proRataAPR);
    emissionSources.push(valueSourceToCachedValueSource(proRataSource, vault, tokenEmission(token)));
  }
  return emissionSources;
}

export function valueSourceToCachedValueSource(
  valueSource: ValueSource,
  vault: VaultDefinitionModel,
  type: string,
): CachedValueSource {
  return Object.assign(new CachedValueSource(), {
    addressValueSourceType: `${vault.address}_${type}`,
    address: vault.address,
    type,
    apr: valueSource.apr,
    name: valueSource.name,
    minApr: valueSource.minApr,
    maxApr: valueSource.maxApr,
    boostable: valueSource.boostable,
  });
}

export async function getVaultValueSources(
  chain: Chain,
  vaultDefinition: VaultDefinitionModel,
): Promise<CachedValueSource[]> {
  // manual over ride for removed compounding of vaults - this can be empty
  const NO_COMPOUND_VAULTS = new Set([TOKENS.BREMBADGER, TOKENS.BVECVX, TOKENS.BCVX]);

  let sources: CachedValueSource[] = [];
  try {
    sources = await getVaultPerformance(chain, vaultDefinition);

    const hasNoUnderlying = NO_COMPOUND_VAULTS.has(vaultDefinition.address);
    if (hasNoUnderlying) {
      sources = sources.filter((s) => s.type !== SourceType.Compound && s.type !== SourceType.PreCompound);
    }

    return sources;
  } catch (err) {
    console.log({ vaultDefinition, err, sources });
    return [];
  }
}

export async function getProtocolValueSources(
  chain: Chain,
  vaultDefinition: VaultDefinitionModel,
): Promise<CachedValueSource[]> {
  try {
    switch (vaultDefinition.protocol) {
      case Protocol.Sushiswap:
        return SushiswapStrategy.getValueSources(chain, vaultDefinition);
      case Protocol.Curve:
      case Protocol.Convex:
        return ConvexStrategy.getValueSources(chain, vaultDefinition);
      case Protocol.Uniswap:
        return UniswapStrategy.getValueSources(vaultDefinition);
      case Protocol.Quickswap:
        return QuickswapStrategy.getValueSources(vaultDefinition);
      case Protocol.Swapr:
        return SwaprStrategy.getValueSources(chain, vaultDefinition);
      case Protocol.OxDAO:
        return OxDaoStrategy.getValueSources(chain, vaultDefinition);
      case Protocol.Aura:
      case Protocol.Balancer:
        return BalancerStrategy.getValueSources(vaultDefinition);
      default: {
        return [];
      }
    }
  } catch (error) {
    console.log({ error, message: `Failed to update value sources for ${vaultDefinition.protocol}` });
    return [];
  }
}
