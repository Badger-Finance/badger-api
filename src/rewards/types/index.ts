export { ClaimableBalance } from './claimable-balance';
export { UserClaimMetadata } from './user-claim-metadata';
export { UserClaimSnapshot } from './user-claim-snapshot';
