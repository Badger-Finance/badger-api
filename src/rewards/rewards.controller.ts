import { Controller, Get, Inject, PathParams } from '@tsed/common';
import { ContentType } from '@tsed/schema';
import { AirdropMerkleClaim, RewardMerkleClaim } from '../interface/MerkleDistribution';
import { RewardsService } from './rewards.service';

@Controller('/reward')
export class RewardController {
  @Inject()
  rewardsService!: RewardsService;

  @ContentType('json')
  @Get('/gitcoin/:address')
  async getGitcoinAirdropClaim(@PathParams('address') address: string): Promise<AirdropMerkleClaim> {
    return this.rewardsService.getUserAirdrop('gitcoin-airdrop.json', address);
  }

  @ContentType('json')
  @Get('/tree/:address')
  async getBadgerTreeReward(@PathParams('address') address: string): Promise<RewardMerkleClaim> {
    return this.rewardsService.getUserRewards(address);
  }

  @ContentType('json')
  @Get('/shop/:address')
  async getBadgerShopEntry(@PathParams('address') address: string): Promise<void> {
    return this.rewardsService.checkBadgerShopEligibility(address);
  }
}
