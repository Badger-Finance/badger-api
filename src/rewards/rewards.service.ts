import { Service } from '@tsed/common';
import { NotFound } from '@tsed/exceptions';
import { BigNumber, ethers } from 'ethers';
import { getObject } from '../aws/s3.utils';
import { Chain } from '../chains/config/chain.config';
import { ChainNetwork } from '../chains/enums/chain-network.enum';
import { BOUNCER_PROOFS, ONE_YEAR_SECONDS, REWARD_DATA } from '../config/constants';
import { TOKENS } from '../config/tokens.config';
import { Digg__factory, RewardsLogger__factory } from '../contracts';
import { getPrice } from '../prices/prices.utils';
import { uniformPerformance } from '../protocols/interfaces/performance.interface';
import { createValueSource, ValueSource } from '../protocols/interfaces/value-source.interface';
import { SettDefinition } from '../setts/interfaces/sett-definition.interface';
import { getCachedSett } from '../setts/setts.utils';
import { formatBalance, getToken } from '../tokens/tokens.utils';
import { Boost } from './interfaces/boost.interface';
import { BoostData } from './interfaces/boost-data.interface';
import { BoostMultipliers } from './interfaces/boost-multipliers.interface';
import { Eligibility } from './interfaces/eligibility.interface';
import { AirdropMerkleClaim, AirdropMerkleDistribution } from './interfaces/merkle-distributor.interface';
import { RewardMerkleClaim } from './interfaces/reward-merkle-claim.interface';
import { getTreeDistribution } from './rewards.utils';

@Service()
export class RewardsService {
  /**
   * Get airdrop merkle claim for a user.
   * @param airdrop Airdrop JSON filename.
   * @param address User Ethereum address.
   */
  async getUserAirdrop(airdrop: string, address: string): Promise<AirdropMerkleClaim> {
    const airdropFile = await getObject(REWARD_DATA, airdrop);
    const fileContents: AirdropMerkleDistribution = JSON.parse(airdropFile.toString('utf-8'));
    const claim = fileContents.claims[address.toLowerCase()];
    if (!claim) {
      throw new NotFound(`${address} does not qualify for airdrop`);
    }
    return claim;
  }

  /**
   * Get airdrop merkle claim for a user.
   * @param airdrop Airdrop JSON filename.
   * @param address User Ethereum address.
   */
  async getBouncerProof(address: string): Promise<AirdropMerkleClaim> {
    const airdropFile = await getObject(REWARD_DATA, BOUNCER_PROOFS);
    const fileContents: AirdropMerkleDistribution = JSON.parse(airdropFile.toString('utf-8'));
    const claim = fileContents.claims[ethers.utils.getAddress(address)];
    if (!claim) {
      throw new NotFound(`${address} is not on the bouncer list`);
    }
    return claim;
  }

  /**
   * Get badger tree reward merkle claim for a user.
   * @param address User Ethereum address.
   */
  async getUserRewards(address: string): Promise<RewardMerkleClaim> {
    const treeDistribution = await getTreeDistribution();
    const claim = treeDistribution.claims[address];
    if (!claim) {
      throw new NotFound(`${address} does not have claimable rewards.`);
    }
    return claim;
  }

  /**
   * Get badger shop eligibility for a user.
   * Returns 200 on eligible, 403 on not eligible.
   * @param address User Ethereum address.
   */
  async checkBouncerList(address: string): Promise<Eligibility> {
    let eligible = false;
    try {
      await this.getBouncerProof(address);
      eligible = true;
    } catch (err) {} // not found, not eligible
    return {
      isEligible: eligible,
    };
  }

  static async getUserBoosts(addresses: string[]): Promise<Record<string, Boost>> {
    const boostFile = await getObject(REWARD_DATA, 'badger-boosts.json');
    const fileContents: BoostData = JSON.parse(boostFile.toString('utf-8'));
    const defaultMultipliers: BoostMultipliers = {};
    Object.keys(fileContents.multiplierData).forEach(
      (key) => (defaultMultipliers[key] = fileContents.multiplierData[key].min),
    );
    const userBoosts: Record<string, Boost> = {};
    for (const address of addresses) {
      let boostData = fileContents.userData[address.toLowerCase()];
      if (!boostData) {
        boostData = {
          boost: 1,
          stakeRatio: 0,
          nftMultiplier: 1,
          multipliers: defaultMultipliers,
          nativeBalance: 0,
          nonNativeBalance: 0,
        };
      } else {
        const userMulipliers = boostData.multipliers;
        const missingMultipliers = Object.keys(defaultMultipliers).length !== Object.keys(userMulipliers).length;
        if (missingMultipliers) {
          const includedMultipliers = new Set();
          const totalPercentile = Object.entries(userMulipliers)
            .map((entry) => {
              const [key, value] = entry;
              includedMultipliers.add(key);
              const min = fileContents.multiplierData[key].min;
              const max = fileContents.multiplierData[key].max;
              const range = max - min;
              return (value - min) / range;
            })
            .reduce((total, value) => (total += value), 0);
          const percentile = totalPercentile / Object.entries(userMulipliers).length;
          Object.entries(fileContents.multiplierData).forEach((entry) => {
            const [key, value] = entry;
            if (!includedMultipliers.has(key)) {
              const range = value.max - value.min;
              userMulipliers[key] = value.min + percentile * range;
            }
          });
        }
        boostData.multipliers = userMulipliers;
      }
      userBoosts[address] = boostData;
    }
    return userBoosts;
  }

  static async getRewardEmission(chain: Chain, settDefinition: SettDefinition): Promise<ValueSource[]> {
    if (!chain.rewardsLogger || settDefinition.depositToken === TOKENS.DIGG) {
      return [];
    }
    const { settToken } = settDefinition;
    const sett = await getCachedSett(settDefinition);
    const boostFile = await getObject(REWARD_DATA, 'badger-boosts.json');
    const boostData: BoostData = JSON.parse(boostFile.toString('utf-8'));
    let boostRange = { min: 1, max: 1 };
    if (boostData.multiplierData) {
      boostRange = boostData.multiplierData[sett.vaultToken];
    }

    // create relevant contracts
    const rewardsLogger = RewardsLogger__factory.connect(chain.rewardsLogger, chain.provider);

    let sharesPerFragment = BigNumber.from(1);
    if (chain.network === ChainNetwork.Ethereum) {
      const diggContract = Digg__factory.connect(TOKENS.DIGG, chain.provider);
      sharesPerFragment = await diggContract._sharesPerFragment();
    }

    // filter active unlock schedules
    const unlockSchedules = await rewardsLogger.getAllUnlockSchedulesFor(settToken);

    if (unlockSchedules.length === 0) {
      return [];
    }

    const cutoff = new Date();
    cutoff.setMinutes(-20);
    const cutOffTimestamp = parseInt((cutoff.getTime() / 1000).toString());
    const includedTokens = new Set();
    const activeSchedules = unlockSchedules
      .slice()
      .sort((a, b) => b.end.sub(a.end).toNumber())
      .filter((schedule) => {
        if (includedTokens.has(schedule.token)) {
          return false;
        }
        const isActive = schedule.end.gte(cutOffTimestamp);
        if (isActive) {
          includedTokens.add(schedule.token);
        }
        return isActive;
      });

    /**
     * Calculate rewards emission percentages:
     *   - P: Price of Token
     *   - A: Amount of Token Emitted
     *   - D: Duration of Token Emission in Seconds
     *   - S: Number of Seconds Per Year
     *   - L: Amount of Value Locked in Sett
     *
     * Reward APR is determined by the extrapolated one year value emitted
     * from the unlock schedues as a percentage of the current locked value.
     *
     * APR = (P * A * (S / D)) / L * 100
     *
     * This vaule is calculated for all tokens emitted for any given sett
     * and persisted against that sett. There is a 20 minute grace period for
     * emission that have since ended, and only the latest active emission
     * will be used for yield calcuation.
     */
    const emissionSources: ValueSource[] = [];
    for (const schedule of activeSchedules) {
      const price = await getPrice(schedule.token);
      const token = getToken(schedule.token);
      let emission = schedule.totalAmount;
      if (token.address === TOKENS.DIGG) {
        emission = emission.div(sharesPerFragment);
      }
      const amount = formatBalance(emission, token.decimals);
      const durationScalar = ONE_YEAR_SECONDS / schedule.duration.toNumber();
      const yearlyEmission = price.usd * amount * durationScalar;
      const apr = (yearlyEmission / sett.value) * 100;
      emissionSources.push(createValueSource(`${token.name} Rewards`, uniformPerformance(apr), false, boostRange));
    }
    return emissionSources;
  }
}
