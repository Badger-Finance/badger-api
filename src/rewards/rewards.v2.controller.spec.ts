import { PlatformServerless } from '@tsed/platform-serverless';
import { PlatformServerlessTest } from '@tsed/platform-serverless-testing';

import { NetworkStatus } from '../errors/enums/network-status.enum';
import { setupMockChain } from '../test/mocks.utils';
import { randomClaimSnapshots } from '../test/mocks.utils/mock.helpers';
import { RewardsService } from './rewards.service';
import { RewardsV2Controller } from './rewards.v2.controller';

describe('rewards.v2.controller', () => {
  const claimAmounts = 40;

  beforeEach(
    PlatformServerlessTest.bootstrap(PlatformServerless, {
      lambda: [RewardsV2Controller],
    }),
  );
  afterEach(() => PlatformServerlessTest.reset());

  beforeEach(() => {
    setupMockChain();
    const records = randomClaimSnapshots(claimAmounts);
    jest.spyOn(RewardsService.prototype, 'list').mockImplementation(async () => ({
      count: claimAmounts,
      records,
    }));
  });

  describe('list', () => {
    it('returns page zero with default page size per page with no query params', async () => {
      const { body, statusCode } = await PlatformServerlessTest.request.get('/rewards');
      expect(statusCode).toEqual(NetworkStatus.Success);
      expect(JSON.parse(body)).toMatchSnapshot();
    });
  });
});
