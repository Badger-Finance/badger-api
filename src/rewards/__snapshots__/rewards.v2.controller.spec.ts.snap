// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rewards.v2.controller list returns page zero with default page size per page with no query params 1`] = `
Object {
  "total_count": 40,
  "total_page_num": 2,
  "users": Array [
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0x798D1bE841a82a273720CE31c822C61a67a601C3": 2e-9,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0x798D1bE841a82a273720CE31c822C61a67a601C3": 2e-9,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0x798D1bE841a82a273720CE31c822C61a67a601C3": 2e-9,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0x798D1bE841a82a273720CE31c822C61a67a601C3": 2e-9,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0x798D1bE841a82a273720CE31c822C61a67a601C3": 2e-9,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0xBA485b556399123261a5F9c95d413B4f93107407": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0xBA485b556399123261a5F9c95d413B4f93107407": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0xBA485b556399123261a5F9c95d413B4f93107407": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0xBA485b556399123261a5F9c95d413B4f93107407": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0xBA485b556399123261a5F9c95d413B4f93107407": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0x798D1bE841a82a273720CE31c822C61a67a601C3": 2e-9,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0x798D1bE841a82a273720CE31c822C61a67a601C3": 2e-9,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0x798D1bE841a82a273720CE31c822C61a67a601C3": 2e-9,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0x798D1bE841a82a273720CE31c822C61a67a601C3": 2e-9,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 1,
        "0x798D1bE841a82a273720CE31c822C61a67a601C3": 2e-9,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0xBA485b556399123261a5F9c95d413B4f93107407": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0xBA485b556399123261a5F9c95d413B4f93107407": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0xBA485b556399123261a5F9c95d413B4f93107407": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0xBA485b556399123261a5F9c95d413B4f93107407": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
    Object {
      "rewards": Object {
        "0xBA485b556399123261a5F9c95d413B4f93107407": 1,
      },
      "user_addr": "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28",
    },
  ],
}
`;
